### ISSUE TITLE
Regex quantifier with {} doesn't match the correct number of characters in some cases

### ISSUE THREAD

**Issue by SirBogman** [rakudo/rakudo#5588]
```
say "burden" ~~ /(.**{2})+n/
```

## Expected Behavior

```
｢urden｣
 0 => ｢ur｣
 0 => ｢de｣
```

## Actual Behavior

```
｢burden｣
 0 => ｢bu｣
 0 => ｢rd｣
 0 => ｢e｣
```

## Notes

The following produces the expected behavior. The difference is that it contains `2`, instead of `{2}`.
```
say "burden" ~~ /(.**2)+n/
```

## Environment
* Compiler version (`rakudo -v` or `raku -v`):
* Welcome to Rakudo™ v2023.11.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2023.11.

**Comment by jubilatious1** [comment]
Raku regexes don't require `{}` curlies around quantifiers.
In fact, `{}` curlies are reserved for running code within regex matchers (and associated substitution functions).


```perl6
  ~ % raku
Welcome to Rakudo™ v2023.05.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2023.05.

To exit type 'exit' or '^D'
[0] > say "burden" ~~ /(.**2)+n/
｢urden｣
 0 => ｢ur｣
 0 => ｢de｣
```
Further tests:
```perl6
[0] > say "burden" ~~ /(.**{"elephantiasis"})+n/
Cannot use non-Numeric or NaN value as quantifier
  in regex  at <unknown file> line 1
  in block <unit> at <unknown file> line 1

[0] > say "burden" ~~ /(.**{"2"})+n/
｢burden｣
 0 => ｢bu｣
 0 => ｢rd｣
 0 => ｢e｣
[0] > say "burden" ~~ /(.**{"2".so})+n/
｢burden｣
 0 => ｢b｣
 0 => ｢u｣
 0 => ｢r｣
 0 => ｢d｣
 0 => ｢e｣
```
Well, it seems to understand that the `2` within your `{}` curlies is a number, but it's doing something to the regex engine. 

My best guess is it's screwing up backtracking. Below turns off backtracking with the `:r` ratchet adverb (named argument):
```perl6
[0] > say "burden" ~~ m/(.**{2})+n/
｢burden｣
 0 => ｢bu｣
 0 => ｢rd｣
 0 => ｢e｣
[0] > say "burden" ~~ m:r/(.**{2})+n/
｢urden｣
 0 => ｢ur｣
 0 => ｢de｣
[0] >
```
Possibly related to https://github.com/Raku/problem-solving/issues/390#

**Comment by pmichaud** [comment]
I'm reluctant to say that it completely "understands" the {2} within the curlies:
```
[0] > say "burden" ~~ /(.**{"3"})+n/
｢burden｣
 0 => ｢bur｣
 0 => ｢de｣
[0] > say "burden" ~~ /(.**{"4"})+n/
｢burden｣
 0 => ｢bu｣
 0 => ｢rde｣
[0] > say "burdeburden" ~~ /(.**{4})+n/
｢burdeburden｣
 0 => ｢burd｣
 0 => ｢ebu｣
 0 => ｢rde｣
[0] > say "burdeburden" ~~ /(.**{3})+n/
｢burdeburden｣
 0 => ｢bur｣
 0 => ｢deb｣
 0 => ｢ur｣
 0 => ｢de｣
```

The first example of attempting to match 4 characters is instructive, in that it returns matches where none of them are 4 characters long.   I think there must be something else gong on with the block evaluation of the ```**``` quantifier.

Pm

**Comment by jubilatious1** [comment]
Now this is making some sense to me: the first test uses `[]` grouping, and I get the answer I expect. 
The second test with `()` captures seems to be the culprit?

```perl6
[0] > say "burden" ~~ /[.**{"3"}]+n/
｢rden｣
[0] > say "burden" ~~ /(.**{"3"})+n/
｢burden｣
 0 => ｢bur｣
 0 => ｢de｣
[0] >
```
MoarVM version 2023.05.

**Comment by SirBogman** [comment]
> Raku regexes don't require `{}` curlies around quantifiers.

I know that I don't need the `{}` in my example. But I was trying to keep it simple. In my actual use case, I have a variable. For example:

```
say "burden" ~~ /(.**{$r})+n/
```

Even though I just have a constant in `{}`, it should still work. Can we run the binary searcher to see if this behavior has ever changed?

**Comment by jubilatious1** [comment]
Someone not knowing Raku regexes might read this issue and think `{}` curlies are necessary (similar to Perl). Hence my clarification.
```perl6
[0] > say "burden" ~~ /(.**{1+2})+n/
｢burden｣
 0 => ｢bur｣
 0 => ｢de｣
[0] > say "burden" ~~ /[.**{1+2}]+n/
｢rden｣
[0] > say "burden" ~~ m:r/(.**{1+2})+n/
｢rden｣
 0 => ｢rde｣
[0] > say "burden" ~~ m:r/[.**{1+2}]+n/
｢rden｣
[0] >
[0] > #MoarVM version 2023.05.
Nil
```

Still thinking this is a capture issue that loses count when backtracking is enabled.

Cheers.

**Comment by SirBogman** [comment]
> ```raku
> ```raku
> 
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> Still thinking this is a capture issue that loses count when backtracking is enabled.
> ```

Yes, this seems right to me. I forgot about the non-capturing groups `[]`. That actually works for my use case, but this still seems like a valid issue.

**Comment by raiph** [comment]
cf #4105
